### PR TITLE
Formatting session interfaces

### PR DESCRIPTION
**Pull request by nemakam**
## Description
1. Updated session related interfaces
2. Removed `InnerReceiver` from `MessageSession`.
3. Added docs for the same

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] **I have read the [contribution guidelines](./CONTRIBUTING.md).**
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR.
- [x] The pull request does not introduce breaking changes (unless a major version change occurs in the assembly and module).
- [x] If applicable, the public code is properly documented.
- [x] Pull request includes test coverage for the included changes.
- [x] The code builds without any errors.

### CODE REVIEW

File: src/Microsoft.Azure.ServiceBus/IQueueClient.cs [hunk 1]
```diff
@@ -14,5 +18,27 @@ public interface IQueueClient : IReceiverClient, ISenderClient
         /// Gets the name of the queue.
         /// </summary>
         string QueueName { get; }
+
+        /// <summary>
+        /// Receive session messages continously from the queue. Registers a message handler and begins a new thread to receive session-messages.
```
Comment (vinaysurya) [inline src/Microsoft.Azure.ServiceBus/IQueueClient.cs]: >message [](start = 77, length = 7)

nit: session handler

Comment (vinaysurya) [inline src/Microsoft.Azure.ServiceBus/IQueueClient.cs]: actually its fine

---
In reply to: [124972167](https://github.com/Azure/azure-service-bus-dotnet/pull/202#discussion_r124972167) [](ancestors = 124972167)

File: src/Microsoft.Azure.ServiceBus/ISubscriptionClient.cs [hunk 2]
```diff
@@ -43,5 +46,27 @@ public interface ISubscriptionClient : IReceiverClient
         /// <param name="ruleName">The name of the rule.</param>
         /// <returns>A task instance that represents the asynchronous remove rule operation.</returns>
         Task RemoveRuleAsync(string ruleName);
+
+        /// <summary>
+        /// Receive session messages continously from the queue. Registers a message handler and begins a new thread to receive session-messages.
+        /// This handler(<see cref="Func{T1,T2,T3,TResult}"/>) is awaited on every time a new message is received by the queue client.
```
Comment (vinaysurya) [inline src/Microsoft.Azure.ServiceBus/ISubscriptionClient.cs]: >queue client [](start = 121, length = 12)

nit: subscription client

File: src/Microsoft.Azure.ServiceBus/ISubscriptionClient.cs [hunk 3]
```diff
@@ -43,5 +46,27 @@ public interface ISubscriptionClient : IReceiverClient
...
+        /// <see cref="IMessageSession"/> contains the session information, and must be used to perform Complete/Abandon/Deadletter or other such operations on the <see cref="Message"/></param>
+        /// <param name="exceptionReceivedHandler">A <see cref="Func{T1, TResult}"/> that is invoked during exceptions.
+        /// <see cref="ExceptionReceivedEventArgs"/> contains contextual information regarding the exception.</param>
+        /// <remarks>  Enable prefetch to speeden up the receive rate. 
+        /// Use <see cref="RegisterSessionHandler(Func{IMessageSession,Message,CancellationToken,Task}, SessionHandlerOptions)"/> to configure the settings of the pump.</remarks>
+        void RegisterSessionHandler(Func<IMessageSession, Message, CancellationToken, Task> handler, Func<ExceptionReceivedEventArgs, Task> exceptionReceivedHandler);
+
+        /// <summary>
+        /// Receive session messages continously from the queue. Registers a message handler and begins a new thread to receive session-messages.
+        /// This handler(<see cref="Func{IMessageSession, Message, CancellationToken, Task}"/>) is awaited on every time a new message is received by the queue client.
```
Comment (vinaysurya) [inline src/Microsoft.Azure.ServiceBus/ISubscriptionClient.cs]: >queue client [](start = 154, length = 12)

nit: subscription client

File: src/Microsoft.Azure.ServiceBus/SessionClient.cs [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+        /// <returns>A session object.</returns>
+        public Task<IMessageSession> AcceptMessageSessionAsync()
+        {
+            return this.AcceptMessageSessionAsync(this.ServiceBusConnection.OperationTimeout);
+        }
+
+        /// <summary>
+        /// Gets a session object of any <see cref="IMessageSession.SessionId"/> that can be used to receive messages for that sessionId.
+        /// </summary>
+        /// <param name="serverWaitTime">Amount of time for which the call should wait for to fetch the next session.</param>
```
Comment (vinaysurya) [inline src/Microsoft.Azure.ServiceBus/SessionClient.cs]: >wait for [](start = 82, length = 8)

nit: remove this part.

File: src/Microsoft.Azure.ServiceBus/SessionClient.cs [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+        public Task<IMessageSession> AcceptMessageSessionAsync(string sessionId)
+        {
+            return this.AcceptMessageSessionAsync(sessionId, this.ServiceBusConnection.OperationTimeout);
+        }
+
+        /// <summary>
+        /// Gets a particular session object identified by <paramref name="sessionId"/> that can be used to receive messages for that sessionId.
+        /// </summary>
+        /// <param name="sessionId">The sessionId present in all its messages.</param>
+        /// <param name="serverWaitTime">Amount of time for which the call should wait for to fetch the next session.</param>
```
Comment (vinaysurya) [inline src/Microsoft.Azure.ServiceBus/SessionClient.cs]: >for [](start = 87, length = 3)

nit: remove

File: src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs [hunk 6]
```diff
@@ -1087,7 +1087,8 @@ void ThrowIfNotPeekLockMode()
             }
         }
 
-        void OnMessageHandler(
+        /// <summary> </summary>
+        protected virtual void OnMessageHandler(
```
Comment (vinaysurya) [inline src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs]: >protected virtua [](start = 8, length = 16)

why this change?

File: src/Microsoft.Azure.ServiceBus/SessionReceivePump.cs [hunk 7]
```diff
@@ -121,14 +120,17 @@ Task RaiseExceptionReceived(Exception e, string action)
...
                     await this.maxPendingAcceptSessionsSemaphoreSlim.WaitAsync(this.pumpCancellationToken).ConfigureAwait(false);
-                    session = await this.client.AcceptMessageSessionAsync().ConfigureAwait(false);
+                    IMessageSession session = await this.client.AcceptMessageSessionAsync().ConfigureAwait(false);
                     if (session == null)
                     {
                         await Task.Delay(Constants.NoMessageBackoffTimeSpan, this.pumpCancellationToken).ConfigureAwait(false);
                         continue;
                     }
 
-                    TaskExtensionHelper.Schedule(() => this.MessagePumpTaskAsync(session));
```
Comment (vinaysurya) [inline src/Microsoft.Azure.ServiceBus/SessionReceivePump.cs]: >session [](start = 81, length = 7)

what would have failed without this?

**Comment by vinaysurya** [comment]
:clock1:

File: src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs (hunk 6 above)
Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs]: So that i can override it in MessageSession and throw InvalidOperationException

File: src/Microsoft.Azure.ServiceBus/SessionReceivePump.cs (hunk 7 above)
Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/SessionReceivePump.cs]: In the current code, nothing would have failed since the pump would schedule it before the `session` would be overwritten in the next loop. 
But just changed it to keep it risk free.

File: src/Microsoft.Azure.ServiceBus/MessageSession.cs [hunk 8]
```diff
@@ -1,41 +1,32 @@
...
-        }
-
+        public MessageSession(
+            string entityPath,
+            MessagingEntityType? entityType,
+            ReceiveMode receiveMode,
+            ServiceBusConnection serviceBusConnection,
+            ICbsTokenProvider cbsTokenProvider,
+            RetryPolicy retryPolicy,
+            int prefetchCount = 0,
```
Comment (SeanFeldman) [inline src/Microsoft.Azure.ServiceBus/MessageSession.cs]: Should this be using `Constants`? Similar to how a default is defined for [message pump](https://github.com/Azure/azure-service-bus-dotnet/blob/7c59009fe5b96f533d10d27fc3a24322d386f342/src/Microsoft.Azure.ServiceBus/Constants.cs#L40).
